### PR TITLE
Add Flash forecasting demo

### PR DESCRIPTION
**Pull request by ethanwharris**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: flash_tutorials/electricity_forecasting/.meta.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  - Tabular
+  - Forecasting
+description: |
+  This tutorial covers using Lightning Flash and it's integration with PyTorch Forecasting to train an autoregressive
+  model (NBeats) on hourly electricity pricing data. We show how the built-in interpretability tools from PyTorch
+  Forecasting can be used with Flash to plot the trend and daily seasonality in our data discovered by the model. We
+  also cover how features from PyTorch Lightning such as the learning rate finder can be used easily with Flash. As a
+  bonus, we show hat we can resample daily observations from the data to discover weekly trends instead.
+requirements:
+  - pandas==1.1.5
```
Comment (ethanwharris) [inline flash_tutorials/electricity_forecasting/.meta.yml:16]: Reviewers note: This version is being pinned just for now until an issue with more recent pandas versions is resolved in Flash

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+# Finally, we train the new model:
+
+# %%
+trainer.fit(model, datamodule=datamodule)
+
+# %% [markdown]
+# Now let's look at what it learned:
+
+# %%
+best_model_path = trainer.checkpoint_callback.best_model_path
```
Comment (tchaton) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:270]: Might be worth to make a function of this, so you don't get duplicated code :) Same above.

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+# ## Creating the Flash DataModule
+#
+# Before we can load our data with Flash, we first need to create "time_idx" column.
+# The "time_idx" column should contain integers corresponding to the observation index (e.g. in our case the difference
+# between two "time_idx" values is the number of hours between the observations).
+# To do this we convert the datetime to an index (by taking the nanoseconds value and dividing by the number of
+# nanoseconds in an hour) then subtract the minimum value so it starts at zero:
+
+# %%
+df_energy_hourly["time_idx"] = (df_energy_hourly.index.view(int) / (1e9 * 60 * 60)).astype(int)
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash_tutorials/electricity_forecasting/electricity_forecasting.py:64]: Done :smiley:

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+# The Flash `TabularForecastingData` (which uses the `TimeSeriesDataSet` from PyTorch Forecasting internally) also
+# supports loading data from multiple time series (e.g. you may have electricity data from multiple countries).
+# To indicate that our data is all from the same series, we add a `constant` column with a constant value of zero:
+
+# %%
+df_energy_hourly["constant"] = 0
+
+# %% [markdown]
+# Now we can create our `TabularForecastingData`.
+# We choose two days (48 hours) as the encoder length (`max_encoder_length`) and one day (24 hours) as the prediction
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash_tutorials/electricity_forecasting/electricity_forecasting.py:77]: Added some more discussion here

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    time_idx="time_idx",
+    target="price actual",
+    group_ids=["constant"],
+    max_encoder_length=max_encoder_length,
+    max_prediction_length=max_prediction_length,
+    time_varying_unknown_reals=["price actual"],
+    train_data_frame=df_energy_hourly[lambda x: x.time_idx <= training_cutoff],
+    val_data_frame=df_energy_hourly,
+    batch_size=256,
+)
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash_tutorials/electricity_forecasting/electricity_forecasting.py:96]: I don't think so, ... not sure that works anywhere at the moment

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+# The [PyTorch Forecasting Documentation](https://pytorch-forecasting.readthedocs.io/en/latest/api/pytorch_forecasting.models.nbeats.NBeats.html)
+# recommends `"widths"` of `[32, 512]` but since our data here is quite simple, we halve it to try to prevent
+# overfitting.
+#
+# To understand the `"backcast_loss_ratio"`, take a look at this diagram of the model taken from
+# [the arXiv paper](https://arxiv.org/abs/1905.10437):
+#
+# <center width="100%" style="padding:10px"><img src="diagram.png" width="250px"></center>
+#
+# Each 'block' within the NBeats architecture includes a forecast output and a backcast which can each yield their own
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash_tutorials/electricity_forecasting/electricity_forecasting.py:112]: I think just single quotes is better here, no need to format it as code

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+# %% [markdown]
+# ## Loading the data
+#
+# We'll use the Spanish hourly energy demand generation and weather data set from Kaggle:
+# https://www.kaggle.com/nicholasjhana/energy-consumption-generation-prices-and-weather
+#
+# First, download the data:
+
+# %%
+download_data("https://pl-flash-data.s3.amazonaws.com/kaggle_electricity.zip", "./data")
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:30]: lets use standard path, see
https://github.com/PyTorchLightning/lightning-tutorials/blob/7434905898e25fee8db8c5af97f8abef98518c91/course_UvA-DL/12-meta-learning/Meta_Learning.py#L55

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
+# %% [markdown]
+# In this tutorial we'll look at using [Lightning Flash](https://github.com/PyTorchLightning/lightning-flash) and it's
+# integration with [PyTorch Forecasting](https://github.com/jdb78/pytorch-forecasting) for autoregressive modelling of
+# electricity prices using the N_BEATS model.
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:4]: does the N_BEATS has any reference?

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+# To indicate that our data is all from the same series, we add a `constant` column with a constant value of zero.
+#
+# Here's the full preprocessing function:
+
+# %%
+
+
+def preprocess(df: pd.DataFrame, frequency: str = "1H") -> pd.DataFrame:
+    df["time"] = pd.to_datetime(df["time"], utc=True, infer_datetime_format=True)
+    df = df.set_index("time")
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:65]:
```suggestion
    df.set_index("time", inplace=True)
```

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+training_cutoff = df_energy_hourly["time_idx"].max() - max_prediction_length
+
+datamodule = TabularForecastingData.from_data_frame(
+    time_idx="time_idx",
+    target="price actual",
+    group_ids=["constant"],
+    max_encoder_length=max_encoder_length,
+    max_prediction_length=max_prediction_length,
+    time_varying_unknown_reals=["price actual"],
+    train_data_frame=df_energy_hourly[lambda x: x.time_idx <= training_cutoff],
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:109]: do we need lambda here?

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+# In the [PyTorch Forecasting Documentation](https://pytorch-forecasting.readthedocs.io/en/latest/api/pytorch_forecasting.models.nbeats.NBeats.html),
+# the authors recommend using `"widths"` of `[32, 512]`.
+# In order to prevent overfitting with smaller datasets, a good rule of thumb is to limit the number of parameters of
+# your model.
+# For this reason, we use `"widths"` of `[16, 256]`.
+#
+# To understand the `"backcast_loss_ratio"`, let's take a look at this diagram of the model taken from
+# [the arXiv paper](https://arxiv.org/abs/1905.10437):
+#
+# <center width="100%" style="padding:10px"><img src="diagram.png" width="250px"></center>
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:128]: use MD formatting, then the image is embeded

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+#
+# Tabular models can be particularly sensitive to the choice of learning rate.
+# Helpfully, PyTorch Lightning provides a built-in learning rate finder that suggests a suitable learning rate
+# automatically.
+# Here's how to use it:
+
+# %%
+trainer = flash.Trainer(
+    max_epochs=3,
+    gpus=1 if torch.cuda.is_available() else 0,
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:151]:
```suggestion
    gpus=int(torch.cuda.is_available()),
```

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    gradient_clip_val=0.01,
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:152]: lets talk what or why this is needed

File: flash_tutorials/electricity_forecasting/electricity_forecasting.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    datamodule.parameters,
+    backbone="n_beats",
+    backbone_kwargs={"widths": [16, 256], "backcast_loss_ratio": 1.0},
+    learning_rate=5e-4,
+)
+
+trainer = flash.Trainer(
+    max_epochs=3 * 24,
+    check_val_every_n_epoch=24,
+    gpus=1 if torch.cuda.is_available() else 0,
```
Comment (Borda) [inline flash_tutorials/electricity_forecasting/electricity_forecasting.py:263]:
```suggestion
    gpus=int(torch.cuda.is_available()),
```

**Review by Borda — APPROVED** [review]

**Pull request merged**